### ISSUE TITLE
`include_block block with context` does not include variables defined in templates

### ISSUE THREAD

**Issue by jonnyscholes** [wagtail/wagtail#5279]
Wagtails `include_block` tag is meant to match as closely as possible Jinja2's `include` tag 
- so one would expect that when you include the context the local template variables are also includes. However the following code does not make `foo` available to the template that `block` renders:

```
{% with foo="bar" %}
        {% include_block block with context %}
{% endwith %}
```

Looks like this is because `jinja2.nodes.ContextReference()` does not currently include template defined variables (which we use in [wagtail/core/jinja2tags.py](wagtail/core/jinja2tags.py#L39) to get the context). There is an issue relating [to this on Jinja2's repo](https://github.com/pallets/jinja/issues/860).

We could provide our own Jinja2 environment with a patched `CodeGenerator` fairly easily to include template variables (the method described in the issue above does work and django is nice enough to expose a way to load custom environments).

I'd be keen to submit a PR for this - but I'm unsure if we want to ship a patched `CodeGenerator` class? If we were to do so we'd probably want to pin the Jinja2 version in Wagtail a little more tightly to avoid breakages.

**Comment by jonnyscholes** [comment]
Looks like [there is a PR open](https://github.com/pallets/jinja/pull/861) to add this functionality into the Jinja2 API... hasn't been looked at in a while though.

**Comment by Scotchester** [comment]
Just ran into this myself. Looks like [that Jinja2 PR](https://github.com/pallets/jinja/pull/861) was merged and released in Jinja 2.11.0. @jonnyscholes, if you're still up for contributing here, a PR to update Wagtail's Jinja2 `include_block` tag to support this would be most welcome!
